src/lib.rs

```diff
@@ -136,13 +140,60 @@ pub use crate::params::{ParamParser, ParamValue, Params};
 pub use crate::resultset::{InitWriter, QueryResultWriter, RowWriter, StatementMetaWriter};
 pub use crate::value::{ToMysqlValue, Value, ValueInner};
 
+const SCRAMBLE_SIZE: usize = 20;
+
 /// Implementors of this trait can be used to drive a MySQL-compatible database backend.
 pub trait MysqlShim<W: Write> {
     /// The error type produced by operations on this shim.
     ///
     /// Must implement `From<io::Error>` so that transport-level errors can be lifted.
     type Error: From<io::Error>;
 
+    /// Server version
+    fn version(&self) -> &str {
+        // 5.1.10 because that's what Ruby's ActiveRecord requires
+        "5.1.10-alpha-msql-proxy"
+    }
+
+    /// Connection id
+    fn connect_id(&self) -> u32 {
+        u32::from_le_bytes([0x08, 0x00, 0x00, 0x00])
+    }
+
+    /// get auth plugin
+    fn default_auth_plugin(&self) -> &str {
+        "mysql_native_password"
+    }
+
+    /// get auth plugin
+    fn auth_plugin_for_username(&self, _user: &[u8]) -> &str {
+        "mysql_native_password"
+    }
+
+    /// Default salt(scramble) for auth plugin
+    fn salt(&self) -> [u8; SCRAMBLE_SIZE] {
+        let bs = ";X,po_k}>o6^Wz!/kM}N".as_bytes();
+        let mut scramble: [u8; SCRAMBLE_SIZE] = [0; SCRAMBLE_SIZE];
+        for i in 0..SCRAMBLE_SIZE {
+            scramble[i] = bs[i];
+            if scramble[i] == b'\0' || scramble[i] == b'$' {
+                scramble[i] = scramble[i] + 1;
+            }
+        }
```

Comment:
Where does this code stem from? Do we expect that all implementors must repeat it? If so, we should have it be a part of the library instead.